src/data_migrations/2021-12-14-empty-blocks.ts

```diff
@@ -0,0 +1,146 @@
+/* This is free and unencumbered software released into the public domain. */
+
+import { Config, parseConfig } from '../config.js';
+import { pg, sql } from '../database.js';
+import { emptyTransactionsRoot } from '../utils.js';
+
+import {
+  AccountID,
+  Engine,
+  ConnectEnv,
+  NetworkConfig,
+} from '@aurora-is-near/engine';
+import { program } from 'commander';
+import externalConfig from 'config';
+import pino, { Logger } from 'pino';
+
+const logger = pino(pino.destination(2));
+
+declare global {
+  // eslint-disable-next-line @typescript-eslint/no-namespace
+  namespace NodeJS {
+    // eslint-disable-next-line @typescript-eslint/no-empty-interface
+    interface ProcessEnv extends ConnectEnv {}
+  }
+}
+
+class ReindexWorker {
+  protected readonly contractID: AccountID;
+  protected readonly pgClient: pg.Client;
+
+  constructor(
+    public readonly config: Config,
+    public readonly network: NetworkConfig,
+    public readonly logger: Logger,
+    public readonly engine: Engine
+  ) {
+    this.contractID = AccountID.parse(this.config.engine).unwrap();
+    this.pgClient = new pg.Client(config.database);
+  }
+
+  async run(blockId: number): Promise<void> {
+    await this.pgClient.connect();
+    const step = Number(this.config.batchSize || 1000);
+
+    let startBlockQuery = sql
+      .select('block.id')
+      .from('block')
+      .order('block.id DESC')
+      .limit(1);
+    if (blockId > 0) {
+      startBlockQuery = startBlockQuery.where(
+        sql(`block.id <= $1`, blockId)
+      );
+    }
+    const transactionQueryResult = await this.pgClient.query(
+      startBlockQuery.toParams()
+    );
+    const startBlockId = transactionQueryResult.rows[0].id;
+
+    for (let blockId = startBlockId; blockId > 0; blockId -= step) {
+      const endBlockId = blockId - step > 0 ? blockId - step : 0;
+      logger.info(`Fetching blocks ${endBlockId}..${blockId}`);
+      const query = sql
+        .select('DISTINCT block.id AS block_id')
+        .from('block')
+        .leftJoin('transaction')
+        .on({ 'transaction.block': 'block.id' })
+        .where(
+          sql(
+            `transaction.block IS NULL AND block.id >= $1 AND block.id < $2`,
+            blockId - step,
+            blockId
+          )
+        );
+      const result = await this.pgClient.query(query.toParams());
+      if (result.rows.length > 0) {
+        const blockIds = result.rows.map((row: any) => parseInt(row.block_id));
+        const updateQuery = sql
+          .update('block', { transactions_root: emptyTransactionsRoot() })
+          .where(sql.in('id', blockIds));
```

Comment:
This seems overly complicated. Why not just document an `UPDATE ... WHERE` query in the migration instructions?

Reply:
`UPDATE block
	SET transactions_root = '\x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421'
FROM block b
	LEFT JOIN transaction t ON t.block = b.id
	WHERE t.block IS NULL AND block.id = b.id
`
This would be query, and my concern is how will it perform on an aprox 70 000 000  rows On testnet and 50 000 000 + on mainnet. But I'm gona test locally.

Reply:
Let us stick to batched update to ensure controlled data migration.